builtin/logical/pki/path_manage_issuers.go

```diff
@@ -0,0 +1,254 @@
+
+	// Rather than using certutil.ParsePEMBundle (which restricts the
+	// construction of the PEM bundle), we manually parse the bundle instead.
+	pemBytes := []byte(pemBundle)
+	var pemBlock *pem.Block
+
+	var issuers []string
+	var keys []string
+
+	// By decoding and re-encoding PEM blobs, we can pass strict PEM blobs
+	// to the import functionality (importKeys, importIssuers). This allows
+	// them to validate no duplicate issuers exist (and place greater
+	// restrictions during parsing) but allows this code to accept OpenSSL
+	// parsed chains (with full textual output between PEM entries).
+	for len(bytes.TrimSpace(pemBytes)) > 0 {
+		pemBlock, pemBytes = pem.Decode(pemBytes)
+		if pemBlock == nil {
+			return nil, errutil.UserError{Err: "no data found in PEM block"}
+		}
+
+		pemBlockString := string(pem.EncodeToMemory(pemBlock))
+
+		switch pemBlock.Type {
+		case "CERTIFICATE", "X509 CERTIFICATE":
+			// Must be a certificate
+			issuers = append(issuers, pemBlockString)
+		case "CRL", "X509 CRL":
+			// Ignore any CRL entries.
+		default:
+			// Otherwise, treat them as keys.
+			keys = append(keys, pemBlockString)
+		}
+	}
+
+	if len(keys) > 0 && !keysAllowed {
+		return logical.ErrorResponse("private keys found in the PEM bundle but not allowed by the path; use /issuers/import/bundle"), nil
+	}
+
+	mkc := newManagedKeyContext(ctx, b, req.MountPoint)
+	for keyIndex, keyPem := range keys {
+		// Handle import of private key.
+		key, existing, err := importKeyFromBytes(mkc, req.Storage, keyPem, "")
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error parsing key %v: %v", keyIndex, err)), nil
+		}
+
+		if !existing {
+			createdKeys = append(createdKeys, key.ID.String())
+		}
+	}
+
+	for certIndex, certPem := range issuers {
+		cert, existing, err := importIssuer(mkc, req.Storage, certPem, "")
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error parsing issuer %v: %v\n%v", certIndex, err, certPem)), nil
+		}
+
+		issuerKeyMap[cert.ID.String()] = cert.KeyID.String()
+		if !existing {
+			createdIssuers = append(createdIssuers, cert.ID.String())
+		}
+	}
+
+	response := &logical.Response{
+		Data: map[string]interface{}{
+			"mapping":          issuerKeyMap,
+			"imported_keys":    createdKeys,
+			"imported_issuers": createdIssuers,
+		},
+	}
+
+	if len(createdIssuers) > 0 {
+		err := b.crlBuilder.rebuild(ctx, b, req, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// While we're here, check if we should warn about a bad default key. We
+	// do this unconditionally if the issuer or key was modified, so the admin
+	// is always warned. But if unrelated key material was imported, we do
+	// not warn.
+	config, err := getIssuersConfig(ctx, req.Storage)
+	if err == nil && len(config.DefaultIssuerId) > 0 {
+		// We can use the mapping above to check the issuer mapping.
+		if keyId, ok := issuerKeyMap[string(config.DefaultIssuerId)]; ok && len(keyId) == 0 {
```

Comment:
If `!ok` would we also want to give a warning for missing a key for the default issuer?

Reply:
There's two scenarios where we're not ok:

 1. The config is completely bad. We warn a few other places in this case; [see the origin PR](https://github.com/hashicorp/vault/pull/15178) for context or the `deleteIssuer` code in `storage.go`.
 2. On imports of _unrelated certs/keys_. 

In case #2, we could be importing some other issuer completely unrelated to the configured default. Say, a future root or its key. Here, the import-local `issuerKeyMap` won't contain the default issuer, even if its otherwise perfectly valid.

So, warning on `!ok` alone is not sufficient information by itself.

That's what the comment 4 lines above was trying to convey, let me know if you have other wording suggestions :-)

> But if unrelated key material was imported, we do not warn.